builtin/logical/pki/backend.go

```diff
@@ -377,3 +406,134 @@ func (b *backend) invalidate(ctx context.Context, key string) {
+	// We assume here that these lists are now complete.
+	sort.Slice(b.possibleDoubleCountedSerials, func(i, j int) bool {
+		return b.possibleDoubleCountedSerials[i] < b.possibleDoubleCountedSerials[j]
+	})
+
+	sort.Slice(b.possibleDoubleCountedRevokedSerials, func(i, j int) bool {
+		return b.possibleDoubleCountedRevokedSerials[i] < b.possibleDoubleCountedRevokedSerials[j]
+	})
+
+	listEntriesIndex := 0
+	possibleDoubleCountIndex := 0
+	for true {
+		if listEntriesIndex >= len(entries) {
+			break
+		}
+		if possibleDoubleCountIndex >= len(b.possibleDoubleCountedSerials) {
+			break
+		}
+		if entries[listEntriesIndex] == b.possibleDoubleCountedSerials[possibleDoubleCountIndex] {
+			// This represents a double-counted entry
+			b.decrementTotalCertificatesCount()
+			listEntriesIndex = listEntriesIndex + 1
+			possibleDoubleCountIndex = possibleDoubleCountIndex + 1
+			continue
+		}
+		if entries[listEntriesIndex] < b.possibleDoubleCountedSerials[possibleDoubleCountIndex] {
+			listEntriesIndex = listEntriesIndex + 1
+			continue
+		}
+		if entries[listEntriesIndex] > b.possibleDoubleCountedSerials[possibleDoubleCountIndex] {
+			possibleDoubleCountIndex = possibleDoubleCountIndex + 1
+			continue
+		}
+	}
+
+	listRevokedEntriesIndex := 0
+	possibleRevokedDoubleCountIndex := 0
+	for true {
+		if listRevokedEntriesIndex >= len(revokedEntries) {
+			break
+		}
+		if possibleRevokedDoubleCountIndex >= len(b.possibleDoubleCountedRevokedSerials) {
+			break
+		}
+		if revokedEntries[listRevokedEntriesIndex] == b.possibleDoubleCountedRevokedSerials[possibleRevokedDoubleCountIndex] {
+			// This represents a double-counted revoked entry
+			b.decrementTotalRevokedCertificatesCount()
+			listRevokedEntriesIndex = listRevokedEntriesIndex + 1
+			possibleRevokedDoubleCountIndex = possibleRevokedDoubleCountIndex + 1
+			continue
+		}
+		if revokedEntries[listRevokedEntriesIndex] < b.possibleDoubleCountedRevokedSerials[possibleRevokedDoubleCountIndex] {
+			listRevokedEntriesIndex = listRevokedEntriesIndex + 1
+			continue
+		}
+		if revokedEntries[listRevokedEntriesIndex] > b.possibleDoubleCountedRevokedSerials[possibleRevokedDoubleCountIndex] {
+			possibleRevokedDoubleCountIndex = possibleRevokedDoubleCountIndex + 1
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (b *backend) incrementTotalCertificatesCount(certsCounted bool, newSerial string) {
+	switch {
+	case certsCounted:
+		atomic.AddUint32(b.certCount, 1)
+		metrics.SetGauge([]string{"secrets", "pki", "total_certificates_stored"}, float32(*b.certCount))
+	default:
+		// This is unsafe, but a good best-attempt
+		b.possibleDoubleCountedSerials = append(b.possibleDoubleCountedSerials, newSerial)
+	}
+}
+
+func (b *backend) decrementTotalCertificatesCount() {
+	atomic.AddUint32(b.certCount, ^uint32(0))
+	metrics.SetGauge([]string{"secrets", "pki", "total_certificates_stored"}, float32(*b.certCount))
+}
+
+func (b *backend) incrementTotalRevokedCertificatesCount(certsCounted bool, newSerial string) {
```

Comment:
I'm of two minds here, part of me really wants to get rid of the `certsCounted` boolean parameter here and just embed it within the case as `b.certsCounted` so callers don't have to worry/get it wrong about where this boolean should be coming from. 

I see the point of grabbing it before/after storage calls and passing it in but I support my fear of someone using this badly passing in a false value leading to a mem leak is making me nervous.

Reply:
Walking the middle line:
* I've added a comment on how to use certsCounted.
* I've added a check that the list to append to isn't nil.  This should prevent memory leak.
- I don't know how to make this nil-check atomic, so if you have ideas, or a willing to chat about it, that would be really nice.  I'm hesitant to add a recover() from a panic here, since like, oom should panic.

Reply:
Yeah this is a good middle ground, as we've discussed I believe this week, I think we are okay with the nil-check not being 100% atomic here and we shouldn't have any issues with a panic here as go will create a new array under the hood within the append call. Please correct me if I'm incorrect here!